openpype/scripts/otio_burnin.py

```diff
@@ -263,6 +220,139 @@ def add_timecode(
+        longest_list_len = max(
+            len(item["values"]) for item in listed_keys.values()
+        )
+        # Where to store formatted values per frame by key
+        new_listed_keys = [{} for _ in range(longest_list_len)]
+        # Find the longest value per fill key.
+        #   The longest value is used to determine size of burnin box.
+        longest_value_by_key = {}
+        for key, item in listed_keys.items():
+            values = item["values"]
+            # Fill the missing values from the longest list with the last
+            #   value to make sure all values have same "frame cound"
+            last_value = values[-1] if values else ""
+            for _ in range(longest_list_len - len(values)):
+                values.append(last_value)
+
+            # Prepare dictionary structure for nestes values
+            # - last key is overriden on each frame loop
+            item_keys = list(item["keys"])
+            fill_data = {}
+            sub_value = fill_data
+            last_item_key = item_keys.pop(-1)
+            for item_key in item_keys:
+                sub_value[item_key] = {}
+                sub_value = sub_value[item_key]
+
+            # Fill value per frame
+            key_max_len = 0
+            key_max_value = ""
+            for value, new_values in zip(values, new_listed_keys):
+                sub_value[last_item_key] = value
+                try:
+                    value = key.format(**sub_value)
+                except (TypeError, KeyError, ValueError):
+                    value = MISSING_KEY_VALUE
+                new_values[key] = value
+
+                value_len = len(value)
+                if value_len > key_max_len:
+                    key_max_value = value
+                    key_max_len = value_len
+
+            # Store the longest value
+            longest_value_by_key[key] = key_max_value
+
+        # Make sure the longest value of each key is replaced for text size
+        #   calculation
+        for key, value in longest_value_by_key.items():
+            text_for_size = text_for_size.replace(key, value)
+
+        # Create temp file with instructions for each frame of text
+        with tempfile.NamedTemporaryFile(mode="w", delete=False) as temp:
+            path = temp.name
+            lines = []
+            for frame, value in enumerate(new_listed_keys):
+                seconds = float(frame) / fps
+                # Escape special character
+                new_text = text
+                for _key, _value in value.items():
+                    _value = str(_value)
+                    new_text = new_text.replace(_key, str(_value))
+
+                new_text = (
+                    str(new_text)
+                    .replace("\\", "\\\\")
+                    .replace(",", "\\,")
+                    .replace(":", "\\:")
+                )
+                lines.append(
+                    f"{seconds} drawtext@{align} reinit text='{new_text}';")
```

Comment:
We don't seem to write to the file as we generate the `lines` as such this can be outside of the `with` statement.

It seems we're doing "tempfile + write lines + add to cleanup files" quite a bit.
How about creating a method to replace that:

```python
path = self.as_temp_file("\n".join(lines))
```

```python
def as_temp_file(self, text):
    with tempfile.NamedTemporaryFile(mode="w", delete=False) as temp:
        path = temp.name
        temp.write(text)
    self.cleanup_path.append(path)
    return path
```